File: rtl/src/main/c/blaise_str.c
/*
 * Blaise RTL — string operation functions
 *
 * String representation (shared with blaise_arc.c):
 *   +--[4 bytes]--+--[4 bytes]--+--[4 bytes]--+--[N bytes]--+--[1 byte]--+
 *   | RefCount    | Length      | Capacity    | UTF-8 data  | NUL        |
 *   +-------------+-------------+-------------+-------------+------------+
 *   ^--- string pointer (header ptr)
 *
 * nil (0) represents an empty / unassigned string.
 * RefCount = -1 marks immortal (statically-allocated) strings.
 *
 * All functions that return a new string allocate a fresh header with
 * RefCount = 0 (unowned). The compiler's ARC wrapper calls _StringAddRef
 * at the assignment site, bringing RefCount to 1.
 */

#include <stdint.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <ctype.h>

#define IMMORTAL_REFCNT (-1)

typedef struct {
    int32_t refcnt;
    int32_t length;
    int32_t capacity;
    /* char data[]; follows immediately */
} BlaiseStrHdr;

/* ------------------------------------------------------------------ */
/* Internal helpers                                                     */
/* ------------------------------------------------------------------ */

static inline BlaiseStrHdr* str_hdr(void* ptr) {
    return (BlaiseStrHdr*)ptr;
}

static inline const char* str_data(void* ptr) {
    return ptr ? (const char*)ptr + sizeof(BlaiseStrHdr) : "";
}

static inline int32_t str_len(void* ptr) {
    return ptr ? str_hdr(ptr)->length : 0;
}

/* Allocate a new Blaise string of exactly `len` bytes (plus NUL).
   RefCount is set to 0 (unowned); caller must call _StringAddRef. */
static void* str_alloc(int32_t len) {
    BlaiseStrHdr* h = (BlaiseStrHdr*)malloc(sizeof(BlaiseStrHdr) + len + 1);
    if (!h) return NULL;
    h->refcnt   = 0;
    h->length   = len;
    h->capacity = len;
    ((char*)(h + 1))[len] = '\0';
    return (void*)h;
}

/* ------------------------------------------------------------------ */
/* _StringLength(s) : Integer                                           */
/* ------------------------------------------------------------------ */

int32_t _StringLength(void* s) {
    return str_len(s);
}

/* ------------------------------------------------------------------ */
/* _StringPos(sub, s) : Integer  — 1-based; 0 if not found            */
/* ------------------------------------------------------------------ */

int32_t _StringPos(void* sub, void* s) {
    const char* haystack = str_data(s);
    const char* needle   = str_data(sub);
    int32_t     nlen     = str_len(sub);

    if (nlen == 0) return 1;  /* empty needle is always at position 1 */

    const char* found = strstr(haystack, needle);
    if (!found) return 0;
    return (int32_t)(found - haystack) + 1;  /* convert to 1-based */
}

/* ------------------------------------------------------------------ */
/* _StringCopy(s, from, count) : string  — 1-based from               */
/* ------------------------------------------------------------------ */

void* _StringCopy(void* s, int32_t from, int32_t count) {
    int32_t     slen = str_len(s);
    const char* data = str_data(s);

    /* Clamp to valid range (1-based indexing, Delphi semantics) */
    if (from < 1) from = 1;
    int32_t start = from - 1;  /* 0-based offset */
    if (start >= slen) {
        /* Return empty string */
        return str_alloc(0);
    }
    if (count < 0) count = 0;
    /* Use 64-bit arithmetic to avoid signed overflow when count = MaxInt */
    if ((int64_t)start + (int64_t)count > (int64_t)slen) count = slen - start;

    void* result = str_alloc(count);
    if (result && count > 0)
        memcpy((char*)result + sizeof(BlaiseStrHdr), data + start, count);
    return result;
}

/* ------------------------------------------------------------------ */
/* _StringUpperCase(s) : string                                         */
/* ------------------------------------------------------------------ */

void* _StringUpperCase(void* s) {
    int32_t     len  = str_len(s);
    const char* data = str_data(s);
    void*       r    = str_alloc(len);
    if (!r) return NULL;
    char* dst = (char*)r + sizeof(BlaiseStrHdr);
    for (int32_t i = 0; i < len; i++)
        dst[i] = (char)toupper((unsigned char)data[i]);
    return r;
}

/* ------------------------------------------------------------------ */
/* _StringLowerCase(s) : string                                         */
/* ------------------------------------------------------------------ */

void* _StringLowerCase(void* s) {
    int32_t     len  = str_len(s);
    const char* data = str_data(s);
    void*       r    = str_alloc(len);
    if (!r) return NULL;
    char* dst = (char*)r + sizeof(BlaiseStrHdr);
    for (int32_t i = 0; i < len; i++)
        dst[i] = (char)tolower((unsigned char)data[i]);
    return r;
}

/* ------------------------------------------------------------------ */
/* _StringTrim(s) : string — strip leading and trailing whitespace     */
/* ------------------------------------------------------------------ */

void* _StringTrim(void* s) {
    int32_t     len  = str_len(s);
    const char* data = str_data(s);
    int32_t     lo   = 0;
    int32_t     hi   = len - 1;
    while (lo <= hi && (unsigned char)data[lo]  <= ' ') lo++;
    while (hi >= lo && (unsigned char)data[hi]  <= ' ') hi--;
    int32_t newlen = (hi >= lo) ? (hi - lo + 1) : 0;
    void*   r      = str_alloc(newlen);
    if (!r) return NULL;
    if (newlen > 0)
        memcpy((char*)r + sizeof(BlaiseStrHdr), data + lo, newlen);
    return r;
}

/* ------------------------------------------------------------------ */
/* _StringSameText(s1, s2) : Boolean (0 or 1)                          */
/* ------------------------------------------------------------------ */

int32_t _StringSameText(void* s1, void* s2) {
    int32_t     len1 = str_len(s1);
    int32_t     len2 = str_len(s2);
    const char* d1   = str_data(s1);
    const char* d2   = str_data(s2);
    int32_t     i;

    if (len1 != len2) return 0;
    for (i = 0; i < len1; i++) {
        if (tolower((unsigned char)d1[i]) != tolower((unsigned char)d2[i]))
            return 0;
    }
    return 1;
}

/* ------------------------------------------------------------------ */
/* _IntToStr(n) : string                                                */
/* ------------------------------------------------------------------ */

void* _IntToStr(int32_t n) {
    char buf[24];
    int  written = snprintf(buf, sizeof(buf), "%d", n);
    if (written < 0) written = 0;
    void* r = str_alloc(written);
    if (r && written > 0)
        memcpy((char*)r + sizeof(BlaiseStrHdr), buf, written);
    return r;
}

/* ------------------------------------------------------------------ */
/* _Int64ToStr(n) : string                                             */
/* ------------------------------------------------------------------ */

void* _Int64ToStr(int64_t n) {
    char buf[24];
    int  written = snprintf(buf, sizeof(buf), "%lld", (long long)n);
    if (written < 0) written = 0;
    void* r = str_alloc(written);
    if (r && written > 0)
        memcpy((char*)r + sizeof(BlaiseStrHdr), buf, written);
    return r;
}

/* ------------------------------------------------------------------ */
/* _StrToInt(s) : Integer                                               */
/* ------------------------------------------------------------------ */

int32_t _StrToInt(void* s) {
    const char* data = str_data(s);
    return (int32_t)strtol(data, NULL, 10);
}

/* ------------------------------------------------------------------ */
/* _StrToInt64(s) : Int64                                               */
/* ------------------------------------------------------------------ */

int64_t _StrToInt64(void* s) {
    const char* data = str_data(s);
    return (int64_t)strtoll(data, NULL, 10);
}

/* ------------------------------------------------------------------ */
/* _OrdAt(s, i) : Integer  — ordinal (ASCII) value of char at         */
/* 1-based position i.  Returns 0 if out of range.                    */
/* ------------------------------------------------------------------ */

int32_t _OrdAt(void* s, int32_t i) {
    int32_t     len  = str_len(s);
    const char* data = str_data(s);
    if (i < 1 || i > len) return 0;
    return (int32_t)(unsigned char)data[i - 1];
}

/* ------------------------------------------------------------------ */
/* _Chr(n) : string  — one-character string holding byte value n.     */
/* Returns an unowned header (RefCount = 0); caller AddRefs.          */
/* ------------------------------------------------------------------ */

void* _Chr(int32_t n) {
    BlaiseStrHdr* h = (BlaiseStrHdr*)str_alloc(1);
    if (!h) return NULL;
    ((char*)(h + 1))[0] = (char)(unsigned char)n;
    return (void*)h;
}

/* ------------------------------------------------------------------ */
/* _StringCompare(s1, s2) : Integer  — case-sensitive (like strcmp)   */
/* ------------------------------------------------------------------ */

int32_t _StringCompare(void* s1, void* s2) {
    const char* d1 = str_data(s1);
    const char* d2 = str_data(s2);
    while (*d1 && (*d1 == *d2)) { d1++; d2++; }
    return (int32_t)((unsigned char)*d1 - (unsigned char)*d2);
}

/* ------------------------------------------------------------------ */
/* _StringCompareText(s1, s2) : Integer  — case-insensitive           */
/* ------------------------------------------------------------------ */

int32_t _StringCompareText(void* s1, void* s2) {
    const char* d1 = str_data(s1);
    const char* d2 = str_data(s2);
    int c1, c2;
    while (*d1) {
        c1 = tolower((unsigned char)*d1);
        c2 = tolower((unsigned char)*d2);
        if (c1 != c2) return (int32_t)(c1 - c2);
        d1++; d2++;
    }
    return (int32_t)(0 - tolower((unsigned char)*d2));
}

/* ------------------------------------------------------------------ */
/* _StringFormat(fmt, ...) : string                                     */
/*                                                                      */
/* Variadic: after the format string pointer, each format argument      */
/* is passed as a (tag: int, value: int or void*) pair:                 */
/*   tag=0 → int32 value (matched to %d in format)                     */
/*   tag=1 → void* Blaise string (matched to %s in format)             */
/*                                                                      */
/* Scans the format string for %d and %s specifiers, consuming one     */
/* (tag, value) pair per specifier.  %% is emitted as a literal %.     */
/* ------------------------------------------------------------------ */

void* _StringFormat(void* fmt, ...) {
    const char* f    = str_data(fmt);
    int         flen = str_len(fmt);

    /* Two-pass: compute output length, then fill buffer. */
    va_list ap1, ap2;
    va_start(ap1, fmt);
    va_copy(ap2, ap1);

    /* Pass 1: compute total output length */
    size_t out_len = 0;
    const char* p  = f;
    const char* end = f + flen;
    while (p < end) {
        if (*p == '%' && p + 1 < end) {
            p++;
            if (*p == 'd') {
                int tag = va_arg(ap1, int);
                if (tag == 0) {
                    int32_t v = (int32_t)va_arg(ap1, int);
                    char tmp[24];
                    out_len += (size_t)snprintf(tmp, sizeof(tmp), "%d", v);
                } else {
                    void* sv = va_arg(ap1, void*);
                    out_len += (size_t)str_len(sv);
                }
                p++;
            } else if (*p == 's') {
                int tag = va_arg(ap1, int);
                if (tag == 1) {
                    void* sv = va_arg(ap1, void*);
                    out_len += (size_t)str_len(sv);
                } else {
                    int32_t v = (int32_t)va_arg(ap1, int);
                    char tmp[24];
                    out_len += (size_t)snprintf(tmp, sizeof(tmp), "%d", v);
                }
                p++;
            } else if (*p == '%') {
                out_len++;
                p++;
            } else {
                out_len++;  /* unknown %x: emit as-is */
                out_len++;
                p++;
            }
        } else {
            out_len++;
            p++;
        }
    }
    va_end(ap1);

    /* Allocate result string */
    void* result = str_alloc((int32_t)out_len);
    if (!result) { va_end(ap2); return NULL; }
    char* dst = (char*)result + sizeof(BlaiseStrHdr);

    /* Pass 2: fill buffer */
    p = f;
    while (p < end) {
        if (*p == '%' && p + 1 < end) {
            p++;
            if (*p == 'd') {
                int tag = va_arg(ap2, int);
                int32_t v;
                if (tag == 0)
                    v = (int32_t)va_arg(ap2, int);
                else {
                    void* sv = va_arg(ap2, void*);
                    v = 0; (void)sv;
                }
                dst += sprintf(dst, "%d", v);
                p++;
            } else if (*p == 's') {
                int tag = va_arg(ap2, int);
                if (tag == 1) {
                    void* sv = va_arg(ap2, void*);
                    int32_t slen = str_len(sv);
                    if (slen > 0)
                        memcpy(dst, str_data(sv), (size_t)slen);
                    dst += slen;
                } else {
                    int32_t v = (int32_t)va_arg(ap2, int);
                    dst += sprintf(dst, "%d", v);
                }
                p++;
            } else if (*p == '%') {
                *dst++ = '%';
                p++;
            } else {
                *dst++ = '%';
                *dst++ = *p++;
            }
        } else {
            *dst++ = *p++;
        }
    }
    va_end(ap2);
    *dst = '\0';
    return result;
}

/* _StringFromPChar(p) : string
 * Converts a NUL-terminated C string into a Blaise ARC string.
 * Returns a new unowned string (RefCount = 0); the compiler's ARC
 * wrapper calls _StringAddRef at the assignment site. */
void* _StringFromPChar(const char* p) {
    int32_t len;
    void*   r;
    if (!p) return str_alloc(0);
    len = (int32_t)strlen(p);
    r   = str_alloc(len);
    if (r && len > 0)
        memcpy((char*)r + sizeof(BlaiseStrHdr), p, (size_t)len);
    return r;
}

File: rtl/src/main/pascal/blaise_str.pas
{
  Blaise - An Object Pascal Compiler
  Copyright (c) 2026 Graeme Geldenhuys
  SPDX-License-Identifier: Apache-2.0 WITH Swift-exception
  Licensed under the Apache License v2.0 with Runtime Library Exception.
  See LICENSE file in the project root for full license terms.
}

{
  Blaise RTL — string operation functions (Pascal port of blaise_str.c)

  String memory layout — data-pointer convention:

    Allocation:
      +--[4 bytes]--+--[4 bytes]--+--[4 bytes]--+--[N bytes]--+--[1 byte]--+
      | RefCount    | Length      | Capacity    | UTF-8 data  | NUL        |
      +-------------+-------------+-------------+-------------+------------+
      ^--- base (raw malloc result)             ^--- DATA POINTER (stored in var)

  The variable slot holds the DATA POINTER (pointing at the first character
  byte).  The 12-byte header lives immediately before it at negative offsets:

      data_ptr − 12  RefCount  (Integer, 4 bytes)
      data_ptr −  8  Length    (Integer, 4 bytes)
      data_ptr −  4  Capacity  (Integer, 4 bytes)
      data_ptr +  0  char data (Length bytes + NUL terminator)

  nil represents an empty / unassigned string.
  RefCount = -1 marks immortal (statically-allocated string literals).
}

unit blaise_str;

{$mode objfpc}{$H+}

interface

{ ------------------------------------------------------------------ }
{ Libc bindings (malloc/free only — all other libc calls replaced)    }
{ ------------------------------------------------------------------ }
function  _libc_malloc(Size: Int64): Pointer; external name 'malloc';
procedure _libc_free(Ptr: Pointer);           external name 'free';

{ ------------------------------------------------------------------ }
{ String RTL public interface                                          }
{ ------------------------------------------------------------------ }
function  _StringLength(S: Pointer): Integer;
function  _StringPos(Sub, S: Pointer): Integer;
function  _StringCopy(S: Pointer; From, Count: Integer): Pointer;
function  _StringUpperCase(S: Pointer): Pointer;
function  _StringLowerCase(S: Pointer): Pointer;
function  _StringTrim(S: Pointer): Pointer;
function  _StringSameText(S1, S2: Pointer): Integer;
function  _IntToStr(N: Integer): Pointer;
function  _Int64ToStr(N: Int64): Pointer;
function  _StrToInt(S: Pointer): Integer;
function  _StrToInt64(S: Pointer): Int64;
function  _OrdAt(S: Pointer; I: Integer): Integer;
function  _Chr(N: Integer): Pointer;
function  _UpCase(N: Integer): Pointer;
function  _StringCompare(S1, S2: Pointer): Integer;
function  _StringCompareText(S1, S2: Pointer): Integer;
function  _StringFromPChar(P: PChar): Pointer;

{ _StringFormat is variadic and remains implemented in blaise_str.c }

implementation

const
  HDR_SIZE = 12;  { 3 x 4-byte integers: RefCount, Length, Capacity }

{ ------------------------------------------------------------------ }
{ Pure-Pascal memory helpers (no libc)                                 }
{ ------------------------------------------------------------------ }

procedure MemCopy(Dst, Src: Pointer; N: Integer);
var
  D, S: PChar;
  I:    Integer;
begin
  D := PChar(Dst);
  S := PChar(Src);
  for I := 0 to N - 1 do
    D[I] := S[I];
end;

{ ------------------------------------------------------------------ }
{ Internal helpers                                                     }
{ ------------------------------------------------------------------ }

{ StrLen: Ptr is the DATA pointer; length lives at Ptr-8 }
function StrLen(Ptr: Pointer): Integer;
var
  P: ^Integer;
begin
  if Ptr = nil then
    Result := 0
  else
  begin
    P := Ptr - 8;   { Length field at data_ptr − 8 }
    Result := P^;
  end;
end;

{ StrData: data IS the pointer — identity for data-pointer convention }
function StrData(Ptr: Pointer): PChar;
begin
  Result := PChar(Ptr);
end;

{ Allocate a new Blaise string of exactly Len bytes plus NUL.
  RefCount = 0 (unowned); caller must call _StringAddRef.
  Returns DATA POINTER (= base + HDR_SIZE). }
function StrAlloc(Len: Integer): Pointer;
var
  TotalL: Int64;
  Base:   Pointer;
  RC, LN, CP: ^Integer;
  NulPtr: PChar;
begin
  TotalL := Int64(HDR_SIZE) + Int64(Len) + 1;
  Base   := _libc_malloc(TotalL);
  if Base = nil then
  begin
    Result := nil;
    Exit;
  end;
  RC  := Base;        { RefCount at base+0 }
  RC^ := 0;
  LN  := Base + 4;    { Length at base+4 }
  LN^ := Len;
  CP  := Base + 8;    { Capacity at base+8 }
  CP^ := Len;
  Result := Base + HDR_SIZE;   { DATA POINTER }
  NulPtr        := PChar(Result);
  NulPtr[Len]   := 0;          { NUL terminator at data+Len }
end;

{ ------------------------------------------------------------------ }
{ _StringLength                                                        }
{ ------------------------------------------------------------------ }

function _StringLength(S: Pointer): Integer;
begin
  Result := StrLen(S);
end;

{ ------------------------------------------------------------------ }
{ _StringPos(Sub, S) : Integer  — 1-based; 0 if not found            }
{ ------------------------------------------------------------------ }

function _StringPos(Sub, S: Pointer): Integer;
var
  SLen, SubLen: Integer;
  SData, SubData: PChar;
  I, J: Integer;
  Match: Boolean;
begin
  SLen   := StrLen(S);
  SubLen := StrLen(Sub);
  if SubLen = 0 then
  begin
    Result := 1;
    Exit;
  end;
  if SubLen > SLen then
  begin
    Result := 0;
    Exit;
  end;
  SData   := StrData(S);
  SubData := StrData(Sub);
  I := 0;
  while I <= SLen - SubLen do
  begin
    Match := True;
    J := 0;
    while J < SubLen do
    begin
      if SData[I + J] <> SubData[J] then
      begin
        Match := False;
        Break;
      end;
      Inc(J);
    end;
    if Match then
    begin
      Result := I + 1;
      Exit;
    end;
    Inc(I);
  end;
  Result := 0;
end;

{ ------------------------------------------------------------------ }
{ _StringCopy(S, From, Count) : string  — 1-based From               }
{ ------------------------------------------------------------------ }

function _StringCopy(S: Pointer; From, Count: Integer): Pointer;
var
  SLen:  Integer;
  Data:  PChar;
  Start: Integer;
begin
  SLen := StrLen(S);
  if From < 1 then From := 1;
  Start := From - 1;
  if Start >= SLen then
  begin
    Result := StrAlloc(0);
    Exit;
  end;
  { Avoid signed overflow when Count is large (Copy(S, N, MaxInt) idiom) —
    compare Count against (SLen - Start) instead of (Start + Count) > SLen. }
  if (Count < 0) or (Count > SLen - Start) then
    Count := SLen - Start;
  Result := StrAlloc(Count);
  if (Result <> nil) and (Count > 0) then
  begin
    Data := StrData(S);
    MemCopy(StrData(Result), Data + Start, Count);
  end;
end;

{ ------------------------------------------------------------------ }
{ _StringUpperCase                                                     }
{ ------------------------------------------------------------------ }

function _StringUpperCase(S: Pointer): Pointer;
var
  Len:     Integer;
  SrcData: PChar;
  DstData: PChar;
  I, C:    Integer;
begin
  Len    := StrLen(S);
  Result := StrAlloc(Len);
  if Result = nil then Exit;
  SrcData := StrData(S);
  DstData := StrData(Result);
  for I := 0 to Len - 1 do
  begin
    C := SrcData[I];
    if (C >= 97) and (C <= 122) then
      DstData[I] := C - 32
    else
      DstData[I] := C;
  end;
end;

{ ------------------------------------------------------------------ }
{ _StringLowerCase                                                     }
{ ------------------------------------------------------------------ }

function _StringLowerCase(S: Pointer): Pointer;
var
  Len:     Integer;
  SrcData: PChar;
  DstData: PChar;
  I, C:    Integer;
begin
  Len    := StrLen(S);
  Result := StrAlloc(Len);
  if Result = nil then Exit;
  SrcData := StrData(S);
  DstData := StrData(Result);
  for I := 0 to Len - 1 do
  begin
    C := SrcData[I];
    if (C >= 65) and (C <= 90) then
      DstData[I] := C + 32
    else
      DstData[I] := C;
  end;
end;

{ ------------------------------------------------------------------ }
{ _StringTrim                                                          }
{ ------------------------------------------------------------------ }

function _StringTrim(S: Pointer): Pointer;
var
  Len:    Integer;
  Data:   PChar;
  Lo, Hi: Integer;
  NewLen: Integer;
begin
  Len  := StrLen(S);
  Data := StrData(S);
  Lo := 0;
  while Lo < Len do
  begin
    if Data[Lo] > 32 then Break;
    Inc(Lo);
  end;
  Hi := Len - 1;
  while Hi >= Lo do
  begin
    if Data[Hi] > 32 then Break;
    Dec(Hi);
  end;
  if Hi >= Lo then
    NewLen := Hi - Lo + 1
  else
    NewLen := 0;
  Result := StrAlloc(NewLen);
  if (Result = nil) or (NewLen = 0) then Exit;
  MemCopy(StrData(Result), Data + Lo, NewLen);
end;

{ ------------------------------------------------------------------ }
{ _StringSameText                                                      }
{ ------------------------------------------------------------------ }

function _StringSameText(S1, S2: Pointer): Integer;
var
  Len1, Len2, I: Integer;
  D1, D2:        PChar;
  C1, C2:        Integer;
begin
  Len1 := StrLen(S1);
  Len2 := StrLen(S2);
  if Len1 <> Len2 then
  begin
    Result := 0;
    Exit;
  end;
  D1 := StrData(S1);
  D2 := StrData(S2);
  for I := 0 to Len1 - 1 do
  begin
    C1 := D1[I];
    C2 := D2[I];
    if (C1 >= 65) and (C1 <= 90) then C1 := C1 + 32;
    if (C2 >= 65) and (C2 <= 90) then C2 := C2 + 32;
    if C1 <> C2 then
    begin
      Result := 0;
      Exit;
    end;
  end;
  Result := 1;
end;

{ ------------------------------------------------------------------ }
{ WriteDecimal — write N as decimal digits into Buf (no NUL).        }
{ Buf must have at least 21 bytes.  Returns character count.         }
{ ------------------------------------------------------------------ }

function WriteDecimal(N: Int64; Buf: PChar): Integer;
var
  IsNeg: Boolean;
  AbsN:  Int64;
  Pos:   Integer;
  Tmp:   Pointer;
  TP:    PChar;
  I, J:  Integer;
  Digit: Integer;
begin
  Tmp := _libc_malloc(22);
  TP  := PChar(Tmp);
  IsNeg := N < 0;
  if IsNeg then
    AbsN := -N
  else
    AbsN := N;
  if AbsN = 0 then
  begin
    TP[0] := 48;  { '0' }
    Pos   := 1;
  end
  else
  begin
    Pos := 0;
    while AbsN > 0 do
    begin
      Digit  := Integer(AbsN mod 10);
      TP[Pos] := 48 + Digit;
      AbsN   := AbsN div 10;
      Inc(Pos);
    end;
  end;
  I := 0;
  if IsNeg then
  begin
    Buf[0] := 45;  { '-' }
    I := 1;
  end;
  J := Pos - 1;
  while J >= 0 do
  begin
    Buf[I] := TP[J];
    Inc(I);
    Dec(J);
  end;
  _libc_free(Tmp);
  Result := I;
end;

{ ------------------------------------------------------------------ }
{ _IntToStr                                                            }
{ ------------------------------------------------------------------ }

function _IntToStr(N: Integer): Pointer;
var
  Buf:     Pointer;
  BP:      PChar;
  Written: Integer;
begin
  Buf     := _libc_malloc(22);
  BP      := PChar(Buf);
  Written := WriteDecimal(Int64(N), BP);
  Result  := StrAlloc(Written);
  if (Result <> nil) and (Written > 0) then
    MemCopy(StrData(Result), Buf, Written);
  _libc_free(Buf);
end;

{ ------------------------------------------------------------------ }
{ _Int64ToStr                                                          }
{ ------------------------------------------------------------------ }

function _Int64ToStr(N: Int64): Pointer;
var
  Buf:     Pointer;
  BP:      PChar;
  Written: Integer;
begin
  Buf     := _libc_malloc(22);
  BP      := PChar(Buf);
  Written := WriteDecimal(N, BP);
  Result  := StrAlloc(Written);
  if (Result <> nil) and (Written > 0) then
    MemCopy(StrData(Result), Buf, Written);
  _libc_free(Buf);
end;

{ ------------------------------------------------------------------ }
{ _StrToInt / _StrToInt64                                             }
{ ------------------------------------------------------------------ }

function _StrToInt(S: Pointer): Integer;
var
  Data:  PChar;
  Neg:   Boolean;
  I, C:  Integer;
  Value: Int64;
begin
  Data  := StrData(S);
  I     := 0;
  Neg   := False;
  if Data[0] = 45 then  { '-' }
  begin
    Neg := True;
    I   := 1;
  end
  else if Data[0] = 43 then  { '+' }
    I := 1;
  Value := 0;
  if Data[I] = 36 then  { '$' — hexadecimal }
  begin
    Inc(I);
    C := Data[I];
    while ((C >= 48) and (C <= 57))
       or ((C >= 65) and (C <= 70))
       or ((C >= 97) and (C <= 102)) do
    begin
      if C >= 97 then
        Value := Value * 16 + Int64(C - 87)   { 'a'..'f' → 10..15 }
      else if C >= 65 then
        Value := Value * 16 + Int64(C - 55)   { 'A'..'F' → 10..15 }
      else
        Value := Value * 16 + Int64(C - 48);  { '0'..'9' }
      Inc(I);
      C := Data[I];
    end;
  end
  else
  begin
    C := Data[I];
    while (C >= 48) and (C <= 57) do
    begin
      Value := Value * 10 + Int64(C - 48);
      Inc(I);
      C := Data[I];
    end;
  end;
  if Neg then
    Result := Integer(-Value)
  else
    Result := Integer(Value);
end;

function _StrToInt64(S: Pointer): Int64;
var
  Data:  PChar;
  Neg:   Boolean;
  I, C:  Integer;
begin
  Data  := StrData(S);
  I     := 0;
  Neg   := False;
  if Data[0] = 45 then
  begin
    Neg := True;
    I   := 1;
  end
  else if Data[0] = 43 then
    I := 1;
  Result := 0;
  if Data[I] = 36 then  { '$' — hexadecimal }
  begin
    Inc(I);
    C := Data[I];
    while ((C >= 48) and (C <= 57))
       or ((C >= 65) and (C <= 70))
       or ((C >= 97) and (C <= 102)) do
    begin
      if C >= 97 then
        Result := Result * 16 + Int64(C - 87)
      else if C >= 65 then
        Result := Result * 16 + Int64(C - 55)
      else
        Result := Result * 16 + Int64(C - 48);
      Inc(I);
      C := Data[I];
    end;
  end
  else
  begin
    C := Data[I];
    while (C >= 48) and (C <= 57) do
    begin
      Result := Result * 10 + Int64(C - 48);
      Inc(I);
      C := Data[I];
    end;
  end;
  if Neg then
    Result := -Result;
end;

{ ------------------------------------------------------------------ }
{ _OrdAt                                                               }
{ ------------------------------------------------------------------ }

function _OrdAt(S: Pointer; I: Integer): Integer;
var
  Len:  Integer;
  Data: PChar;
begin
  Len  := StrLen(S);
  if (I < 1) or (I > Len) then
  begin
    Result := 0;
    Exit;
  end;
  Data   := StrData(S);
  Result := Data[I - 1];
end;

{ ------------------------------------------------------------------ }
{ _Chr                                                                 }
{ ------------------------------------------------------------------ }

function _Chr(N: Integer): Pointer;
var
  DP: PChar;
begin
  Result := StrAlloc(1);
  if Result = nil then Exit;
  DP    := StrData(Result);
  DP[0] := N;
end;

{ ------------------------------------------------------------------ }
{ _UpCase                                                              }
{ ------------------------------------------------------------------ }

function _UpCase(N: Integer): Pointer;
var
  C: Integer;
begin
  C := N;
  if (C >= 97) and (C <= 122) then
    C := C - 32;
  Result := _Chr(C);
end;

{ ------------------------------------------------------------------ }
{ _StringCompare                                                       }
{ ------------------------------------------------------------------ }

function _StringCompare(S1, S2: Pointer): Integer;
var
  Len1, Len2, Len: Integer;
  D1, D2:          PChar;
  I, C1, C2:       Integer;
begin
  Len1 := StrLen(S1);
  Len2 := StrLen(S2);
  if Len1 < Len2 then Len := Len1 else Len := Len2;
  D1 := StrData(S1);
  D2 := StrData(S2);
  for I := 0 to Len - 1 do
  begin
    C1 := D1[I];
    C2 := D2[I];
    if C1 <> C2 then
    begin
      Result := C1 - C2;
      Exit;
    end;
  end;
  Result := Len1 - Len2;
end;

{ ------------------------------------------------------------------ }
{ _StringCompareText                                                   }
{ ------------------------------------------------------------------ }

function _StringCompareText(S1, S2: Pointer): Integer;
var
  Len1, Len2, Len: Integer;
  D1, D2:          PChar;
  I, C1, C2:       Integer;
begin
  Len1 := StrLen(S1);
  Len2 := StrLen(S2);
  if Len1 < Len2 then Len := Len1 else Len := Len2;
  D1 := StrData(S1);
  D2 := StrData(S2);
  for I := 0 to Len - 1 do
  begin
    C1 := D1[I];
    C2 := D2[I];
    if (C1 >= 65) and (C1 <= 90) then C1 := C1 + 32;
    if (C2 >= 65) and (C2 <= 90) then C2 := C2 + 32;
    if C1 <> C2 then
    begin
      Result := C1 - C2;
      Exit;
    end;
  end;
  Result := Len1 - Len2;
end;

{ ------------------------------------------------------------------ }
{ _StringFromPChar                                                     }
{ ------------------------------------------------------------------ }

function _StringFromPChar(P: PChar): Pointer;
var
  Len: Integer;
begin
  if P = nil then
  begin
    Result := StrAlloc(0);
    Exit;
  end;
  Len := 0;
  while P[Len] <> 0 do
    Inc(Len);
  Result := StrAlloc(Len);
  if (Result <> nil) and (Len > 0) then
    MemCopy(StrData(Result), P, Len);
end;

end.
